### ISSUE TITLE
Using Cocoapods installs the "React" pod, creating conflicts

### ISSUE THREAD

**Issue by hhff** [react-native-community/react-native-svg#621]
I'm not sure how to best handle this, but my project uses pods, and when I install react-native-svg as a pod, it depends on the React pod, and installs it.

My builds then start failing because there are 2 versions of React declared.

How are folks dealing with this?

Edit: It's similar to this issue: https://github.com/oblador/react-native-vector-icons/issues/303

**Comment by msand** [comment]
I've been using an ugly hack in the PodFile: https://github.com/invertase/react-native-firebase/issues/414#issuecomment-331128610
```
  post_install do |installer_representation|
      installer_representation.pods_project.targets.each do |target|
          if target.name == "React"
              target.remove_from_project
          end
      end
  end
```
I really wish there was some better way. Please find one for me 😄

**Comment by ianoshorty** [comment]
Having the same issue, but the proposed fix above doesn't work for me. If I add it, I end up with missing component errors.

**Comment by msand** [comment]
You'll also need to link react-native manually, by adding the projects to libraries in your main project, and to your "Linked Frameworks and Libraries" in the Build Phases. Such that you have it linked both manually and using cocoapods, and then add the hack to remove the cocoapods one, such that you're able to archive your project.
https://github.com/facebook/react-native/issues/15838#issuecomment-333455341

**Comment by ianoshorty** [comment]
Thanks for the tip @msand - fixed for me! This definitely needs a less "hacky" fix.

**Comment by msand** [comment]
Great, happy I could help. I managed to get a slightly less hacky version working today. Removed the manually linked version, but then building/running/archiving from XCode breaks. But I found that compiling manually from the cli still works, so I added to npm scripts to package.json
https://github.com/msand/SVGPodTest/commit/5343af586edd31a3547d215f99f893f4e3f88e8d#diff-b9cfc7f2cdf78a7f4b91a753d10865a2R8
```
+    "release": "cd ios && xcodebuild -workspace SVGPodTest.xcworkspace -configuration Release -scheme SVGPodTest -destination id=711F49EC-5747-4393-94DB-77B97D466541 -derivedDataPath build",
+    "archive": "cd ios && xcodebuild -workspace SVGPodTest.xcworkspace -configuration Release -scheme SVGPodTest -derivedDataPath build  -archivePath ./YourApp.xcarchive archive"
```
It would be nice to be able to build from XCode as well. Not sure how its possible that it works from the cli but is broken there.

**Comment by jgresham** [comment]
Having this issues as well. Currently, is the best solution to just no use Cocoapods? Thanks

**Comment by msand** [comment]
Another approach I found is to remove all the manually linked libraries and let cocoapods link React as well. This doesn't allow starting debug mode from xcode though, but works with run-ios in the cli. If you need to debug the obj-c code, I recommend duplicating the target which you're installing the pods in to, and manually link react there. If you can, I would recommend avoiding cocoapods, as it's just another thing which can break in your system. If npm, react-native link and normal xcode works for you, then just use that. If you need something from outside npm, like with react-native-fbsdk (or want swift code, I think) etc., then you need cocoapods at least.

**Comment by petrogad** [comment]
I'm with @jgresham is the best solution to just avoid Cocopods? This is super frustrating that causes duplicate compilations.

**Comment by skurfuerst** [comment]
Hey everybody,

I've investigated the installation of React Native dependencies a little more in-depth; and I've come to a quite clean cocoapod-based solution (I hope). I've summed up the approach in a blog post at https://sandstorm.de/de/blog/post/react-native-managing-native-dependencies-using-xcode-and-cocoapods.html .

Maybe it helps somebody :)

All the best,
Sebastian

**Comment by roycedot** [comment]
Hello,

I've followed the steps outlined by @msand

1) I have the React .xcodeproj files added to Libraries of my main Xcode project
2) libReact.a and libRCTWebSocket.a in "Linked Frameworks and Libraries" on General tab and "Link Binary With Libraries" on Build Phases tab
3) Podfile "postinstall" hack

But, I also had to 

1) edit Other Linker Flags in my main Xcode Build Settings as the $(inherited) portion was still referencing the React framework from Pods.

Otherwise, linking fails.

2) Edit [CP] Embed Pods Frameworks in Build Phases on my main Xcode project to remove React from Input Files and Output Files

3) And you may need to edit ios/Pods/Target Support Files/Pods-<your proj>-frameworks.sh to remove React from the install_framework calls at the bottom. I didn't have to do this last step on a clean project. But did have to on a project where I applied the Podfile fix after trying a bunch of different other stuff.

**Comment by saggiyogesh** [comment]
Fixed conflicts with these changes in `Podfile`

```
pod 'React', :path => '../node_modules/react-native', :subspecs => [
    'Core',
    'CxxBridge',
    'DevSupport',
    # the following ones are the ones taken from "Libraries" in Xcode:
    'RCTAnimation',
    'RCTActionSheet',
    'RCTBlob',
    'RCTGeolocation',
    'RCTImage',
    'RCTLinkingIOS',
    'RCTNetwork',
    'RCTSettings',
    'RCTText',
    'RCTVibration',
    'RCTWebSocket'
  ]
  
  pod 'yoga', :path => '../node_modules/react-native/ReactCommon/yoga'
  pod 'DoubleConversion', :podspec => '../node_modules/react-native/third-party-podspecs/DoubleConversion.podspec'
  pod 'Folly', :podspec => '../node_modules/react-native/third-party-podspecs/Folly.podspec'
  pod 'glog', :podspec => '../node_modules/react-native/third-party-podspecs/GLog.podspec'
  

pod 'RNSVG', :path => '../node_modules/react-native-svg'

post_install do |installer_representation|
    installer_representation.pods_project.targets.each do |target|
        if target.name == "React"
            target.remove_from_project
        end
    end
end
```

**Comment by rjruizes** [comment]
As @roycedot mentioned above, I had to edit Build Settings -> Other Linker Flags.

I removed `$(inherited)`, which was including `-framework "React"` and other frameworks. I added back the other frameworks:`-framework "yoga" -framework "glog" -framework "folly" -framework "DoubleConversion"`

After that I got a successful build and a working debug menu.

**Comment by brianfoody** [comment]
@roycedot sorry a few months later asking you this but could you provide a snippet of how you implemented point (2) above?

Edit [CP] Embed Pods Frameworks in Build Phases on my main Xcode project to remove React from Input Files and Output Files

**Comment by chrisheninger** [comment]
Came across this thread debugging a separate issue– but I recently ran into this same problem and wanted to share in case it helps others.

It's a known issue that CocoaPods + React-Native is a current pain point. If you do a search in this thread for "cocoapods"– you'll see there's a lot of good discussion:

https://github.com/react-native-community/discussions-and-proposals/issues/104

tl;dr– Comment by Facebook: We are currently working on recommending CocoaPods by default for iOS projects and with it we are looking to simplify link to cause fewer problems for people.

If you notice that the react@0.11.0 pod is being installed when you run pod install– you'll want to update your podfile to include React and it's dependencies– in the end your podfile should look similar to this:

https://facebook.github.io/react-native/docs/integration-with-existing-apps#configuring-cocoapods-dependencies

**Comment by tap349** [comment]
> Came across this thread debugging a separate issue– but I recently ran into this same problem and wanted to share in case it helps others.
> 
> It's a known issue that CocoaPods + React-Native is a current pain point. If you do a search in this thread for "cocoapods"– you'll see there's a lot of good discussion:
> 
> [react-native-community/discussions-and-proposals#104](https://github.com/react-native-community/discussions-and-proposals/issues/104)
> 
> tl;dr– Comment by Facebook: We are currently working on recommending CocoaPods by default for iOS projects and with it we are looking to simplify link to cause fewer problems for people.
> 
> If you notice that the react@0.11.0 pod is being installed when you run pod install– you'll want to update your podfile to include React and it's dependencies– in the end your podfile should look similar to this:
> 
> https://facebook.github.io/react-native/docs/integration-with-existing-apps#configuring-cocoapods-dependencies

I just want to add that it might be necessary to remove `React` pod from _ios/Pods/_ directory manually to get rid of naming collision - `pod install` doesn't remove unused pods:

```sh
$ rm -rf ios/Pods/React
```
